### PR TITLE
Added tests for issue #264 and fixed the issue itself.

### PR DESCRIPTION
**Pull request by chedatomasz**
Also Refactored tests to be runnable by pytest, added proper skipping for testing optional dependencies

### CODE REVIEW

**Comment by chedatomasz** [comment]
I believe F401 deserves a skip (or if there is a good way to check if an import is available without actually importing - please fix and inform me )

**Comment by ita9naiwa** [comment]
You've changed a lot!
I'll check this when I have time.(maybe weekend)

File: implicit/bpr.pyx [hunk 1]
```diff
@@ -221,8 +223,10 @@ class BayesianPersonalizedRanking(MatrixFactorizationBase):
                                                                self.verify_negative_samples)
                 progress.update(1)
                 total = len(user_items.data)
-                progress.set_postfix({"correct": "%.2f%%" % (100.0 * correct / (total - skipped)),
-                                      "skipped": "%.2f%%" % (100.0 * skipped / total)})
+                if total!=0 and total!=skipped:
```
Comment (benfred) [inline implicit/bpr.pyx]: nit: Should have spaces around the '!=' operator here for pep8

File: tests/approximate_als_test.py [hunk 2]
```diff
@@ -66,7 +78,7 @@ def test_similar_items(self):
                 # this causes the test_similar_items call to fail if we set regularization to 0
                 self.__regularization = 1.0
                 try:
-                    super(FaissALSGPUTest, self).test_similar_items()
+                    super(RecommenderBaseMixin).test_similar_items()
```
[the remark this answers is not in the document]
Comment (chedatomasz) [reply tests/approximate_als_test.py]: No, in fact it does not. Will fix in the proper PR

File: tests/bpr_test.py [hunk 3]
```diff
@@ -1,23 +1,34 @@
...
-class BPRTest(unittest.TestCase, TestRecommenderBaseMixin):
+class BPR(unittest.TestCase, RecommenderBaseMixin):
 
     def _get_model(self):
         return BayesianPersonalizedRanking(factors=3, regularization=0, use_gpu=False)
 
+    # Test issue #264 causing crashes on empty matrices
+    def test_fit_empty_matrix(self):
+        raw = [[0, 0, 0], [0, 0, 0], [0, 0, 0]]
+        return BayesianPersonalizedRanking().fit(csr_matrix(raw))
```
[the remark this answers is not in the document]
Comment (chedatomasz) [reply tests/bpr_test.py]: Fixed in the proper pr

File: implicit/bpr.pyx (hunk 1 above)
Comment (chedatomasz) [reply implicit/bpr.pyx]: Fixed in the proper PR

**Comment by chedatomasz** [comment]
I have submitted a separate PR for just the fix. When that is merged I will submit another with the test refactor.

**Pull request closed without merging**